### ISSUE TITLE
Explore using freecache for memory storage

### ISSUE THREAD

**Issue by jzelinskie** [chihaya/chihaya#522]
[Freecache](https://github.com/coocood/freecache) could save us lots of custom code in the memory implementation, if it delivers on performance.

In the worst case, we might just be able to borrow some of the optimization ideas from it.

**Comment by polarathene** [comment]
**TL;DR:** Instead of Freecache, you may want to consider looking into [Ristretto](https://github.com/dgraph-io/ristretto) which appears to be a better alternative; or at a higher level (_and potential Redis replacement / alternative_), [Olric](https://github.com/buraksezer/olric).

---

You might prefer to use [Ristretto](https://github.com/dgraph-io/ristretto) instead (released back in 2019), from the devs behind [Badger](https://github.com/dgraph-io/badger)(K/V DB). They have [a pretty decent article on caching in Go that introduces Ristretto](https://dgraph.io/blog/post/introducing-ristretto-high-perf-go-cache/) and compares it to existing cache solutions including Freecache.

They also [wrote a previous article on caching with Go](https://dgraph.io/blog/post/caching-in-go/) that compared Freecache to a few others and identified the state of Go caching, expressing various concerns they had when choosing a cache solution (_they ended up creating Ristretto instead of course_).

---

I am not a Go dev, but was looking into all of this when evaluating to use another Go project designed for HTTP caching called `Souin`. They at an earlier point used Ristretto (_and perhaps still do, I haven't looked into Souin source much_), but made some change to switch to Badger at a later date.

Badger IIRC will keep keys in memory and values on disk. In more recent versions it changed to also keep values under 1MB per key in memory as well (_I can't recall if these also persist to disk, I think they do as that was important for me with the HTTP cache decision_).

Souin also offers an alternative KV store that also can act as an in-memory cache (_it's memory only and can work in a distributed fashion IIRC_) called [`olric`](https://github.com/buraksezer/olric). That's also a Go project that can run embedded within Chihaya. I'm not sure if that would be more desirable for you over what a lower level package like Ristretto would offer, but perhaps it would be more suitable for Chihaya (_could replace Redis for HA?_).

You can find the [author of Olric compare his project to Ristretto here](https://github.com/buraksezer/olric/issues/79#issuecomment-817294185):

> `Olric` is a distributed, embeddable data structure server. It currently implements distributed map and distributed topic data structures. Because of its consistency model, `Olric` is useful as a distributed cache library for your Go applications. It also supports a few eviction policies including LRU.
> 
> I have a little info about `Ristretto` library of DGraph. It is not distributed. So `Olric` is completely different and it can be comparable with `groupcache`.
